File: python/beeai_framework/adapters/a2a/serve/agent_executor.py
# Copyright 2025 © BeeAI a Series of LF Projects, LLC
# SPDX-License-Identifier: Apache-2.0

from typing_extensions import TypeVar, override

from beeai_framework.adapters.a2a.agents._utils import convert_a2a_to_framework_message
from beeai_framework.agents.experimental.events import RequirementAgentStartEvent, RequirementAgentSuccessEvent
from beeai_framework.agents.react import ReActAgentUpdateEvent
from beeai_framework.agents.react.types import ReActAgentIterationResult
from beeai_framework.agents.tool_calling import ToolCallingAgentStartEvent, ToolCallingAgentSuccessEvent
from beeai_framework.backend import AssistantMessage, MessageToolCallContent, ToolMessage
from beeai_framework.emitter import Emitter, EventMeta
from beeai_framework.serve import MemoryManager, init_agent_memory
from beeai_framework.utils.cancellation import AbortController
from beeai_framework.utils.cloneable import Cloneable
from beeai_framework.utils.lists import find_index
from beeai_framework.utils.strings import to_json

try:
    import a2a.server as a2a_server
    import a2a.server.agent_execution as a2a_agent_execution
    import a2a.server.tasks as a2a_server_tasks
    import a2a.types as a2a_types
    import a2a.utils as a2a_utils
except ModuleNotFoundError as e:
    raise ModuleNotFoundError(
        "Optional module [a2a] not found.\nRun 'pip install \"beeai-framework[a2a]\"' to install."
    ) from e

from beeai_framework.agents import AnyAgent
from beeai_framework.backend.message import (
    AnyMessage,
)
from beeai_framework.logger import Logger

AnyAgentLike = TypeVar("AnyAgentLike", bound=AnyAgent, default=AnyAgent)

logger = Logger(__name__)


class BaseA2AAgentExecutor(a2a_agent_execution.AgentExecutor):
    def __init__(
        self,
        agent: AnyAgentLike,
        agent_card: a2a_types.AgentCard,
        *,
        memory_manager: MemoryManager,
        send_trajectory: bool | None = True,
    ) -> None:
        super().__init__()
        self._agent = agent
        self.agent_card = agent_card
        self._abort_controller = AbortController()
        self._memory_manager = memory_manager
        self._send_trajectory = send_trajectory

    @override
    async def execute(
        self,
        context: a2a_agent_execution.RequestContext,
        event_queue: a2a_server.events.EventQueue,
    ) -> None:
        updater = a2a_server_tasks.TaskUpdater(event_queue, context.task_id, context.context_id)  # type: ignore[arg-type]
        if not context.current_task:
            if not context.message:
                raise ValueError("No message found in the request context.")
            context.current_task = a2a_utils.new_task(context.message)
            await updater.submit()

        cloned_agent = await self._agent.clone() if isinstance(self._agent, Cloneable) else self._agent
        await init_agent_memory(cloned_agent, self._memory_manager, context.context_id)
        new_messages = _extract_request_messages(context)

        await updater.start_work()
        try:
            response = await cloned_agent.run(new_messages, signal=self._abort_controller.signal).observe(
                lambda emitter: self._process_events(emitter, context, updater) if self._send_trajectory else ...
            )

            await updater.complete(
                a2a_utils.new_agent_text_message(
                    response.last_message.text,
                    context.context_id,
                    context.task_id,
                )
            )

        except Exception as e:
            logger.exception("Exception during execution")
            await updater.failed(
                message=a2a_utils.new_agent_text_message(str(e)),
            )

    @override
    async def cancel(
        self,
        context: a2a_agent_execution.RequestContext,
        event_queue: a2a_server.events.EventQueue,
    ) -> None:
        self._abort_controller.abort()

    async def _process_events(
        self, emitter: Emitter, context: a2a_agent_execution.RequestContext, updater: a2a_server_tasks.TaskUpdater
    ) -> None:
        pass


class ReActAgentExecutor(BaseA2AAgentExecutor):
    @override
    async def _process_events(
        self,
        emitter: Emitter,
        context: a2a_agent_execution.RequestContext,
        updater: a2a_server_tasks.TaskUpdater,
    ) -> None:
        async def process_event(data: ReActAgentUpdateEvent, event: EventMeta) -> None:
            text = ""
            if isinstance(data.data, ReActAgentIterationResult):
                if data.data.final_answer:
                    text = data.data.final_answer
                elif data.data.tool_output:
                    text = data.data.tool_output
                elif data.data.tool_name or data.data.tool_input:
                    text = to_json({"tool_name": data.data.tool_name, "tool_input": data.data.tool_input})
                elif data.data.thought:
                    text = data.data.thought

            await updater.start_work(
                a2a_utils.new_agent_text_message(
                    text,
                    context.context_id,
                    context.task_id,
                )
                if isinstance(data.data, ReActAgentIterationResult)
                else a2a_utils.new_agent_parts_message(parts=[a2a_types.Part(root=a2a_types.DataPart(data=data.data))]),
            )

        emitter.on("update", process_event)


class ToolCallingAgentExecutor(BaseA2AAgentExecutor):
    @override
    async def _process_events(
        self,
        emitter: Emitter,
        context: a2a_agent_execution.RequestContext,
        updater: a2a_server_tasks.TaskUpdater,
    ) -> None:
        last_msg = None

        async def process_event(
            data: RequirementAgentStartEvent
            | RequirementAgentSuccessEvent
            | ToolCallingAgentStartEvent
            | ToolCallingAgentSuccessEvent,
            event: EventMeta,
        ) -> None:
            nonlocal last_msg
            messages = data.state.memory.messages
            if last_msg is None:
                last_msg = messages[-1]

            cur_index = find_index(messages, lambda msg: msg is last_msg, fallback=-1, reverse_traversal=True)
            for message in messages[cur_index + 1 :]:
                if (isinstance(message, ToolMessage) and message.content[0].tool_name == "final_answer") or (
                    isinstance(message, AssistantMessage)
                    and isinstance(message.content[0], MessageToolCallContent)
                    and message.content[0].tool_name == "final_answer"
                ):
                    continue

                await updater.start_work(
                    a2a_utils.new_agent_parts_message(
                        parts=[
                            a2a_types.Part(
                                root=a2a_types.TextPart(text=content)
                                if isinstance(content, str)
                                else a2a_types.DataPart(data=content.model_dump())
                            )
                            for content in message.content
                        ]
                    ),
                )
                last_msg = message

        emitter.on("start", process_event)
        emitter.on("success", process_event)


def _extract_request_messages(context: a2a_agent_execution.RequestContext) -> list[AnyMessage]:
    return [
        convert_a2a_to_framework_message(message)
        for message in (
            context.current_task.history
            if context.current_task and context.current_task.history
            else [context.message]
            if context.message
            else []
        )
    ]

File: typescript/src/adapters/a2a/serve/agent_executor.ts
/**
 * Copyright 2025 © BeeAI a Series of LF Projects, LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { v4 as uuidv4 } from "uuid";

import { AnyAgent } from "@/agents/types.js";
import { AgentExecutor, RequestContext, ExecutionEventBus } from "@a2a-js/sdk/server";
import { Message, TaskStatusUpdateEvent } from "@a2a-js/sdk";
import { Logger } from "@/logger/logger.js";
import { convertA2AMessageToFrameworkMessage } from "@/adapters/a2a/agents/utils.js";
import { Callback, Emitter } from "@/emitter/emitter.js";
import { ToolCallingAgentCallbacks, ToolCallingAgentRunState } from "@/agents/toolCalling/types.js";
import { ReActAgentCallbacks } from "@/agents/react/types.js";
import { Message as FrameworkMessage } from "@/backend/message.js";

const logger = Logger.root.child({
  name: "A2A server",
});

export abstract class BaseA2AAgentExecutor implements AgentExecutor {
  protected readonly abortControllers = new Map<string, [string, AbortController]>();

  constructor(protected agent: AnyAgent) {}

  async execute(requestContext: RequestContext, eventBus: ExecutionEventBus): Promise<void> {
    const userMessage = requestContext.userMessage;
    const existingTask = requestContext.task || {
      kind: "task",
      id: uuidv4(),
      contextId: userMessage.contextId || uuidv4(),
      status: {
        state: "submitted",
        timestamp: new Date().toISOString(),
      },
      history: [userMessage],
      metadata: userMessage.metadata,
    };

    const taskId = existingTask.id;
    const contextId = existingTask.contextId;

    // Publish initial Task event if it's a new task
    eventBus.publish(existingTask);

    // Publish "working" status update
    const workingStatusUpdate: TaskStatusUpdateEvent = {
      kind: "status-update",
      taskId: taskId,
      contextId: contextId,
      status: {
        state: "working",
        timestamp: new Date().toISOString(),
      },
      final: false,
    };
    eventBus.publish(workingStatusUpdate);

    const abortController = new AbortController();
    this.abortControllers.set(taskId, [contextId, abortController]);

    const agent = await this.agent.clone();
    agent.memory = await agent.memory.clone();
    agent.memory.reset();
    await agent.memory.addMany(
      (existingTask.history || [requestContext.userMessage]).map(
        convertA2AMessageToFrameworkMessage,
      ) || [],
    );

    try {
      // run the agent
      const response = await agent
        .run(
          {},
          {
            signal: abortController.signal,
          },
        )
        .observe(async (emitter) => {
          await this.process_events(emitter, requestContext, eventBus);
        });

      const agentMessage: Message = {
        kind: "message",
        role: "agent",
        messageId: uuidv4(),
        parts: [{ kind: "text", text: response.result.text }],
        taskId: taskId,
        contextId: contextId,
      };

      // Append agent message to task history
      if (!existingTask.history) {
        existingTask.history = [];
      }
      existingTask.history.push(agentMessage);
      eventBus.publish(existingTask);

      // Publish completed status update with agent message
      const finalUpdate: TaskStatusUpdateEvent = {
        kind: "status-update",
        taskId: taskId,
        contextId: contextId,
        status: {
          state: "completed",
          message: agentMessage,
          timestamp: new Date().toISOString(),
        },
        final: true,
      };
      eventBus.publish(finalUpdate);

      eventBus.finished();
    } catch (error) {
      logger.error("Agent execution error:", { error });
      const errorMessage = error instanceof Error ? error.message : String(error);
      // Publish failed status update
      const errorUpdate: TaskStatusUpdateEvent = {
        kind: "status-update",
        taskId: taskId,
        contextId: contextId,
        status: {
          state: "failed",
          message: {
            kind: "message",
            role: "agent",
            messageId: uuidv4(),
            parts: [{ kind: "text", text: `Agent error: ${errorMessage}` }],
            taskId: taskId,
            contextId: contextId,
          },
          timestamp: new Date().toISOString(),
        },
        final: true,
      };
      eventBus.publish(errorUpdate);
      eventBus.finished();
    }
  }

  async process_events(
    _emitter: Emitter,
    _requestContext: RequestContext,
    _eventBus: ExecutionEventBus,
  ): Promise<void> {
    return;
  }

  public cancelTask = async (taskId: string, eventBus: ExecutionEventBus): Promise<void> => {
    if (this.abortControllers.has(taskId)) {
      const [contextId, abortController] = this.abortControllers.get(taskId)!;
      const cancelledUpdate: TaskStatusUpdateEvent = {
        kind: "status-update",
        taskId: taskId,
        contextId: contextId,
        status: {
          state: "canceled",
          timestamp: new Date().toISOString(),
        },
        final: true,
      };
      eventBus.publish(cancelledUpdate);
      abortController.abort();
    }
  };
}

export class ToolCallingAgentExecutor extends BaseA2AAgentExecutor {
  async process_events(
    emitter: Emitter<ToolCallingAgentCallbacks>,
    requestContext: RequestContext,
    eventBus: ExecutionEventBus,
  ): Promise<void> {
    let lastMsg: FrameworkMessage | undefined = undefined;

    const processEvent: Callback<{ state: ToolCallingAgentRunState }> = async ({ state }) => {
      const messages = state.memory.messages;
      if (lastMsg === undefined) {
        lastMsg = messages.at(-1);
      }

      const index = lastMsg ? messages.lastIndexOf(lastMsg) : -1;
      for (const message of messages.slice(index + 1)) {
        const update: TaskStatusUpdateEvent = {
          kind: "status-update",
          taskId: requestContext.taskId,
          contextId: requestContext.contextId,
          status: {
            state: "working",
            message: {
              kind: "message",
              role: "agent",
              messageId: uuidv4(),
              parts: [{ kind: "text", text: JSON.stringify(message.toPlain()) }],
              taskId: requestContext.taskId,
              contextId: requestContext.contextId,
            },
            timestamp: new Date().toISOString(),
          },
          final: false,
        };
        eventBus.publish(update);
        lastMsg = message;
      }
    };
    emitter.on("start", processEvent);
    emitter.on("success", processEvent);
  }
}

export class ReActAgentExecutor extends BaseA2AAgentExecutor {
  async process_events(
    emitter: Emitter<ReActAgentCallbacks>,
    requestContext: RequestContext,
    eventBus: ExecutionEventBus,
  ): Promise<void> {
    emitter.on("update", async ({ update }) => {
      const updateEvent: TaskStatusUpdateEvent = {
        kind: "status-update",
        taskId: requestContext.taskId,
        contextId: requestContext.contextId,
        status: {
          state: "working",
          message: {
            kind: "message",
            role: "agent",
            messageId: uuidv4(),
            parts: [{ kind: "text", text: update.value }],
            taskId: requestContext.taskId,
            contextId: requestContext.contextId,
          },
          timestamp: new Date().toISOString(),
        },
        final: false,
      };
      eventBus.publish(updateEvent);
    });
  }
}
